src/platform-aws.c

```diff
@@ -759,63 +759,141 @@ static int get_rail_vf_idx(struct fi_info *info)
+	 * assume topo ordering is sufficient */
+	if ((num_rails / num_groups) <= 1) {
 		return;
 	}
 
-	for (size_t i = 0; i < num_rails; ++i) {
-		sorted_info_array[i] = NULL;
+	info_array = (struct fi_info **)calloc(num_rails, sizeof(struct fi_info*));
+	if (info_array == NULL) {
+		NCCL_OFI_WARN("Did not reorder arrays due to calloc failure");
+		goto cleanup;
 	}
 
-	size_t rail_map[2] = {0, 2};
+	vf_array = (size_t *)calloc(num_rails, sizeof(size_t));
+	if (vf_array == NULL) {
+		NCCL_OFI_WARN("Did not reorder arrays due to calloc failure");
+		goto cleanup;
+	}
 
-	for (size_t i = 0; i < num_rails; ++i) {
-		if (info_list_in == NULL) {
-			goto error;
+	/* copy the input list into an array so that we can more *
+	 * easily associate more data (like the vf array) with the
+	 * input and keep everything organized */
+	info_iter = *info_list;
+	info_count = 0;
+	while (info_iter != NULL) {
+		info_array[info_count] = fi_dupinfo(info_iter);
+		if (info_array[info_count] == NULL) {
+			NCCL_OFI_WARN("fi_dupinfo failed");
+			goto cleanup;
 		}
+		info_iter = info_iter->next;
 
-		int vf_idx = get_rail_vf_idx(info_list_in);
-		if (vf_idx < 0 || vf_idx >= 2) {
-			NCCL_OFI_WARN("Invalid vf_idx value %d", vf_idx);
-			goto error;
+		int ret = get_rail_vf_idx(info_array[info_count]);
+		if (ret < 0) {
+			NCCL_OFI_WARN("lookup of rail for index %lu failed: %s",
+				      info_count, strerror(-ret));
+			goto cleanup;
+		}
+		vf_array[info_count] = ret;
+		if (vf_array[info_count] > highest_vf_idx) {
+			highest_vf_idx = vf_array[info_count];
 		}
 
-		size_t rail_idx = rail_map[vf_idx];
-		rail_map[vf_idx]++;
+		info_count++;
+	}
+	if (info_count != num_rails) {
+		NCCL_OFI_WARN("Info count (%lu) and num_rails (%lu) do not match.  Aborting reorder.",
+			      info_count, num_rails);
+		goto cleanup;
+	}
 
-		NCCL_OFI_TRACE(NCCL_INIT | NCCL_NET, "Assigning rail index %d to info list idx %d",
-			rail_idx, i);
+	/* No reorder required, as devices all have the same vf idx
+	   and end result would be the input array */
+	if (highest_vf_idx == 0) {
+		goto cleanup;
+	}
 
-		if (sorted_info_array[rail_idx]) {
-			NCCL_OFI_WARN("Attempted to fill rail slot with duplicate infos");
-			goto error;
-		}
-		sorted_info_array[rail_idx] = info_list_in;
+	for (size_t i = 0 ; i < num_rails ; i++) {
+		size_t j = num_rails;
+		for (j = 0 ; j < num_rails ; j++) {
+			if (info_array[j] == NULL) {
+				continue;
+			}
 
-		info_list_in = info_list_in->next;
+			if (vf_array[j] == next_vf_idx) {
+				if (output_info_list == NULL) {
+					output_info_list = output_info_end = info_array[j];
+				} else {
+					output_info_end->next = info_array[j];
+					output_info_end = info_array[j];
+				}
+				info_array[j] = NULL;
+				next_vf_idx = (next_vf_idx + 1) % (highest_vf_idx + 1);
+				break;
+			}
+		}
+		if (j == num_rails) {
+			NCCL_OFI_WARN("Did not find a device with expected index %zu", next_vf_idx);
+			goto cleanup;
+		}
 	}
```

Comment:
I think the libfabric EFA provider should do this unconditionally when providing the info list to apps.

Reply:
we talked about this 1:1 and agreed to do it this way.